ui/sd_internal/task_manager.py

```diff
@@ -341,6 +354,21 @@ def get_cached_task(session_id:str, update_ttl:bool=False):
         return None
     return task_cache.tryGet(session_id)
 
+def get_devices():
```

Comment:
My intention behind a UI-based device selection was to list *all* available devices (GPUs and CPU), not just the ones we picked up from config.json, or created render threads for.

The intent is to write the chosen devices to config.json. I'd like to avoid the user needing to edit config.json manually for device selection.

This kind of UI-device device selection will improve the adoption of this feature (visible in the UI), and make it simpler for users to use this. It's friendlier to be presented with a list of devices available on the PC, and choose one or multiple, and have that be written to config.json automatically for them.

Reply:
It's a start to be able to select one of the present and active device instead of only CPU in the render path.
But it is true that this should be extended to list all device with it's current status (ready/disabled) and allow to start/stop render threads. But this base is needed to implement the rest and should be tested first.

It's just a matter of extending the current method to include more informations and allow more control.
Ill do this over time in another PR.

Reply:
Thanks, agreed. I think we're pretty close. Just need the getDevices() to return the CPU plus loop over `torch.cuda.device_count()` to get the GPU names.

That'll unblock us to make the UI change to select different combinations of devices, and have it be saved into config.json.

This will be useful even if it needs a program restart (to implement quickly, without starting/stopping threads automatically).

Reply:
And apologies for my comment. I'm not expecting or asking you to build the full getDevices() method, and I'm genuinely happy with this start. I just wanted to put my thoughts across on where we need to build this further, not asking you to build it. All of what we do here is totally voluntary.